Bank api inv setup locks

TheHaliax opened the pull request:
added slot/item lockstate coverage in Rs2Bank API
implemented these changes in InventorySetups with rudimentary auto-sort based on exact setup configuration

TheHaliax committed:
Bank API InvSetups Locks

added slot/item lockstate coverage in Rs2Bank API
implemented these changes in InventorySetups with rudimentary auto-sort based on exact setup configuration

TheHaliax committed:
Merge branch 'API' into BankAPI-InvSetupLocks

TheHaliax committed:
Update InventorySetupItemSerializable.java

TheHaliax committed:
Update Runelite.run.xml

gmason0 reviewed (changes requested):
Can we revert some of the formatting changes  that were made here? makes the PR difficult to follow.

TheHaliax committed:
Refactor inventory setups plugin for code style consistency

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/bank/Rs2Bank.java:
```diff
@@ -1447,16 +1447,16 @@ private static Rs2ItemModel findBankItem(String name, boolean exact) {
      */
     @SuppressWarnings("UnnecessaryLocalVariable")
     private static Rs2ItemModel findBankItem(String name, boolean exact, int amount) {
```
Lets clean up the changes here.

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/bank/Rs2Bank.java:
```diff
@@ -2443,4 +2443,187 @@ private static boolean isBankPluginEnabled() {
+        return Rs2Inventory.items().stream()
+                .filter(item -> {
+                    if (exact) return item.getName().equalsIgnoreCase(name);
+                    return item.getName().toLowerCase().contains(name.toLowerCase());
```
We can use a ternary expression here - `.filter(item -> exact ? item.getName().equalsIgnoreCase(name) : item.getName().toLowerCase().contains(name.toLowerCase())`

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/bank/Rs2Bank.java:
```diff
@@ -2443,4 +2443,187 @@ private static boolean isBankPluginEnabled() {
     }
-}
+
+    public static List<Integer> findLockedSlots() {
```
I think we should refactor this a bit as there is some duplicate code here


```java
private static boolean isWidgetLocked(Widget widget) {
    if (widget == null) return false;
    String[] actions = widget.getActions();
    return actions != null && Arrays.stream(actions).filter(Objects::nonNull).anyMatch("Unlock-slot"::equalsIgnoreCase);
}

public static boolean isLockedSlot(int slot) {
    if (slot < 0 || slot >= 28) return false;
    Widget container = Microbot.getClient().getWidget(BANK_INVENTORY_ITEM_CONTAINER);
    if (container == null || container.getChildren() == null || slot >= container.getChildren().length)
        return false;
    return isWidgetLocked(container.getChild(slot));
}

public static List<Integer> findLockedSlots() {
    List<Integer> lockedSlots = new ArrayList<>();
    Widget container = Microbot.getClient().getWidget(BANK_INVENTORY_ITEM_CONTAINER);
    if (container == null || container.getChildren() == null) {
        log.info("Bank inventory container is null or has no children.");
        return lockedSlots;
    }
    Widget[] items = container.getChildren();
    for (int i = 0; i < items.length; i++) {
        Widget w = items[i];
        if (w == null) {
            log.debug("Slot {}: null widget", i);
            continue;
        }
        log.info("Slot {}: actions = {}", i, Arrays.toString(w.getActions()));
        if (isWidgetLocked(w)) {
            log.info("Found locked slot at {}", i);
            lockedSlots.add(i);
        }
    }
    if (lockedSlots.isEmpty()) {
        log.info("No locked slots detected.");
    }
    return lockedSlots;
}
```

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/bank/Rs2Bank.java:
```diff
@@ -2443,4 +2443,187 @@ private static boolean isBankPluginEnabled() {
+        if (!hasItemLockEnabled || !hasItemLockInventoryOptionEnabled) return false;
+        container = BANK_INVENTORY_ITEM_CONTAINER;
+        invokeMenu(10, rs2Item);
+        return onBankInventoryUpdate();
```
Instead of storing the value of varbit within a class member & using this onBankInventoryUpdate() method, I think a better approach would be 

```java
...
final int currentLockState = Microbot.getVarbitValue(VarbitID.BANK_SIDE_SLOT_OVERVIEW);
invokeMenu(10, rs2Item);
return sleepUntilTrue(() -> Microbot.getVarbitValue(VarbitID.BANK_SIDE_SLOT_OVERVIEW) != currentLockState, 300, 2000);
```

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/bank/Rs2Bank.java:
```diff
@@ -2443,4 +2443,187 @@ private static boolean isBankPluginEnabled() {
+            } else {
+                log.info("Failed to unlock item in slot {}", slot);
+            }
+            sleep(200); // Delay helps prevent race conditions or missed UI updates
```
We can remove this sleep when using the approach I recommended above for `toggleItemLock(Rs2ItemModel)`

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/bank/Rs2Bank.java:
```diff
@@ -2443,4 +2443,187 @@ private static boolean isBankPluginEnabled() {
+            } else {
+                log.info("Failed to lock item '{}' in slot {}", item.getName(), item.getSlot());
+            }
+            sleep(200); // Optional: spacing to prevent interface race
```
We can remove this sleep when using the approach I recommended above for `toggleItemLock(Rs2ItemModel)`

gmason0 commented on runelite-client/src/main/java/net/runelite/client/plugins/microbot/util/bank/Rs2Bank.java:
```diff
@@ -2443,4 +2443,187 @@ private static boolean isBankPluginEnabled() {
+            } else {
+                log.info("Failed to lock item '{}' in slot {}", item.getName(), slot);
+            }
+            sleep(200);
```
We can remove this sleep when using the approach I recommended above for `toggleItemLock(Rs2ItemModel)`

gmason0 reviewed (changes requested):
Great feature - lets just clean up some of the requests here. I also would like to see more of these log messages move to log.debug instead of log.info to avoid spamming the client.log file.

TheHaliax committed:
Enhance inventory setup slot and lock handling

Improved inventory setup logic to handle item slot assignments and locking more robustly. Added methods to lock items based on setup, improved detection and toggling of locked slots, and replaced info logs with debug logs for less verbosity. Refactored slot and lock management for better reliability and maintainability.

TheHaliax committed:
Add detailed Javadoc to inventory and bank lock utilities

Added comprehensive Javadoc comments to methods in Rs2InventorySetup and Rs2Bank related to inventory and bank slot locking. These comments clarify method purposes, preconditions, postconditions, and side effects, improving code maintainability and developer understanding. No functional changes were made.

TheHaliax committed:
Add toggleLockOnSlot method to handle item locking

Introduces the toggleLockOnSlot method to toggle the locked state of inventory setup items, update the layout, persist configuration changes, and refresh the UI. Also improves documentation for inventory slot locking logic.

TheHaliax committed:
Merge branch 'development' into BankAPI-InvSetupLocks
